src/snowflake/snowpark/_internal/ast.py

```diff
@@ -0,0 +1,114 @@
+
+
+@dataclass
+class TrackedCallable:
+    var_id: int
+    func: Callable
+
+
+class AstBatch:
+    """
+    A batch of AST statements. This class is used to generate AST requests.
+
+    The core statement types are:
+    - Assign: Creates a new variable and assigns a value to it.
+    - Eval: Evaluates a variable.
+    """
+
+    # Function used to generate request IDs. This is overridden in some tests.
+    generate_request_id = uuid.uuid4
+
+    def __init__(self, session) -> None:
+        """
+        Initializes a new AST batch.
+
+        Args:
+            session: The Snowpark session.
+        """
+        self._session = session
+        self.reset_id_gen()
+        self._init_batch()
+
+        # Track callables in this dict (memory id -> TrackedCallable).
+        self._callables = {}
+
+    def reset_id_gen(self):
+        """Resets the ID generator."""
+        self._id_gen = itertools.count(start=1)
+
+    def assign(self, symbol=None):
+        """
+        Creates a new assignment statement.
+
+        Args:
+            symbol: An optional symbol to name the new variable.
+        """
+        stmt = self._request.body.add()
+        # TODO: extended BindingId spec from the branch snowpark-ir.
+        stmt.assign.uid = next(self._id_gen)
+        stmt.assign.var_id.bitfield1 = stmt.assign.uid
+        stmt.assign.symbol.value = symbol if isinstance(symbol, str) else ""
+        return stmt.assign
+
+    def eval(self, target):
+        """
+        Creates a new evaluation statement.
+
+        Args:
+            target: The variable to evaluate.
+        """
+        stmt = self._request.body.add()
+        stmt.eval.uid = next(self._id_gen)
+        stmt.eval.var_id.CopyFrom(target.var_id)
+
+    def flush(self) -> Tuple[str, str]:
+        """Ties off a batch and starts a new one. Returns the tied-off batch."""
+        batch = str(base64.b64encode(self._request.SerializeToString()), "utf-8")
+        self._init_batch()
+        return (str(self._request_id), batch)
+
+    def _init_batch(self):
+        self._request_id = AstBatch.generate_request_id()  # Generate a new unique ID.
+        self._request = proto.Request()
+
+        (major, minor, patch) = VERSION
+        self._request.client_version.major = major
+        self._request.client_version.minor = minor
+        self._request.client_version.patch = patch
+
+        (major, minor, micro, releaselevel, serial) = sys.version_info
+        self._request.client_language.python_language.version.major = major
+        self._request.client_language.python_language.version.minor = minor
+        self._request.client_language.python_language.version.patch = micro
+        self._request.client_language.python_language.version.label = releaselevel
+
+        self._request.client_ast_version = CLIENT_AST_VERSION
+
+    # TODO(SNOW-1491199) - This method is not covered by tests until the end of phase 0. Drop the pragma when it is covered.
```

Comment:
nit: maybe we can add callable related code later.

Reply:
As I mentioned before, merging back into `main` in a very fine-grained fashion probably doesn't pass a cost/benefit analysis - it's a lot more work to sort the changes this way, and for very limited benefit, since we aim to have all the production code merged within another 5-10 pull requests.